### ISSUE TITLE
proposal: `no_logic_in_state_constructor`

### ISSUE THREAD

**Issue by goderbauer** [dart-lang/linter#3059]
## Description

The constructor of a `State` object should never contain any logic. That logic should always go into `State.initState`. When people put this kind of logic into their constructor, they often run into trouble.

## Kind

Guard against errors.

## Good Examples 
```dart
class FooState extends State<Foo> with SingleTickerProviderStateMixin {
  late AnimationController _controller;

  @override
  void initState() {
  super.initState();
  _controller = AnimationController(vsync: this);
  }
  // ...
}
```
## Bad Examples

```dart
class FooState extends State<Foo> with SingleTickerProviderStateMixin {
  FooState() {
   _controller = AnimationController(vsync: this);
  }

  late AnimationController _controller;

  // ...
}
```

## Discussion

*Add any other motivation or useful context here.*

### Discussion checklist

- [x] List any existing rules this proposal complements, overlaps or conflict with.
- [x] List any relevant issues (reported here, the [SDK Tracker], or elsewhere).
- [x] If there's any prior art (e.g., in other linters), please add references here.
- [x] If this proposal corresponds to [Effective Dart] or [Flutter Style Guide] advice, please call it out. (If there isn’t any corresponding advice, should there be?)
- [x] If this proposal is motivated by real-world examples, please provide as many details as you can.  Demonstrating potential impact is _especially valuable_.

[Writing Lints]: https://github.com/dart-lang/linter/blob/master/doc/WritingLints.MD
[Effective Dart]: https://dart.dev/guides/language/effective-dart
[Flutter Style Guide]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo]
[SDK Tracker]: https://github.com/dart-lang/sdk/issues

/cc @dnfield

**Comment by dnfield** [comment]
I wonder if we can/should also lint that you don't otherwise initialize variables in `State` objects, e.g.

```dart
class FooState extends State<Foo> {
  // Should be initializing this in initState rather than here
  final MySpecialControllerThatDoesNotNeedThis _controller = MySpecialControllerThatDoesNotNeedThis(bar: 3);

  ...
}
```

**Comment by goderbauer** [comment]
That seems overly limiting. It's totally fine (and common) to initialize a `TextEditingController` like that, for example.

**Comment by dnfield** [comment]
Maybe if it's `late final` but if it's `final` you might be initializing an object you'll never use right?

**Comment by bwilkerson** [comment]
What does the word "logic" mean in this context? The proposed implementation flags constructors whose body is not empty, implying that no code should be executed when an instance is created, but that appears to conflict with the claim that it's acceptable to initialize a `TextEditingController` (though if it's only `late final` fields that can have initializers, then that would be consistent with not executing code when the instance is created).

**Comment by minhqdao** [comment]
I don't think there are any conflicts because a `TextEditingController` also shouldn't get initialized in the code block of a constructor.

```dart
class FooState extends State<Foo> with SingleTickerProviderStateMixin {
  FooState() {
    // Never do any of these
    _animationController = AnimationController(vsync: this);
    _textEditingController = TextEditingController();
  }

  late final AnimationController _animationController;
  late final TextEditingController _textEditingController;
  // ...
}
```

What @goderbauer meant was that, instead of initializing in `initState`, it's also totally fine to initialize a `TextEditingController` as a member of a `State` object right away (answer to [this](https://github.com/dart-lang/linter/issues/3059#issuecomment-966504528) comment):

```dart
class FooState extends State<Foo> with SingleTickerProviderStateMixin {
  // Both cool and should not be linted, therefore no conflict
  late final AnimationController _animationController = AnimationController(vsync: this);
  final TextEditingController _textEditingController = TextEditingController();
  // ...
}
```

**Comment by srawlins** [comment]
Is the proposal here that a State constructor should not even have a body? Or is it a set of statements that should be banned, like assignments, method calls, ...
